core/src/lib.rs

```diff
@@ -23,145 +23,41 @@
-//!     .and_then(|socket, _| {
-//!         apply_outbound(socket, Ping::default()).map_err(|e| e.into_io_error())
-//!     })
-//!     // TODO: right now the only available protocol is ping, but we want to replace it with
-//!     //       something that is more simple to use
-//!     .dial("/ip4/127.0.0.1/tcp/12345".parse::<libp2p_core::Multiaddr>().unwrap()).unwrap_or_else(|_| panic!())
-//!     .and_then(|mut pinger| {
-//!         pinger.ping(());
-//!         let f = pinger.into_future().map(|_| ()).map_err(|(e, _)| e);
-//!         Box::new(f) as Box<Future<Item = _, Error = _>>
-//!     });
-//!
-//! // Runs until the ping arrives.
-//! let mut rt = Runtime::new().unwrap();
-//! let _ = rt.block_on(ping_dialer).unwrap();
-//! # }
-//! ```
-//!
-//! ## Grouping protocols
-//!
-//! You can use the `.or_upgrade()` method to group multiple upgrades together. The return value
-//! also implements the `ConnectionUpgrade` trait and will choose one of the protocols amongst the
-//! ones supported.
+//! # General overview
+//!
+//! From a user point of view, the main concepts of libp2p are:
+//!
+//! - A `PeerId` is a unique global identifier for a node on the network. Each node should have a
+//!   different `PeerId`. Normally, a `PeerId` is the hash of the public key used to negotiate
+//!   encryption on the communication channel, thereby guaranteeing that they cannot be spoofed.
+//! - The `Transport` trait define how to reach a remote or listen for incoming remote connections.
+//!   See the `transport` module.
+//! - The `Swarm` struct contains all the active or pending connections to remotes, including all
+//!   the opened protocols.
+//! - The `NetworkBehaviour` trait is implemented on types that can customize the behaviour of the
+//!   `Swarm`. It is the `NetworkBehaviour` that defines what happens on the network.
+//! - The `Topology` trait is implemented on types that can hold the layout of the network.
+//!
+//! In order to get started with libp2p, proceed with the following steps:
+//!
+//! - Create an instance of a type that implements `Transport` based on the list of transport
+//!   protocols (such as TCP/IP, DNS, bluetooth, etc.) that you want to support. See the
+//!   `transport` module for information about how to do that.
+//! - Create an instance of a type that implements `NetworkBehaviour`. See the `swarm` module for
+//!   information about how to do that.
+//! - Create an instance of a swarm that implements `Topology`. See the `topology` module for
+//!   information about how to do that.
+//! - Create a `Swarm` by passing the three objects you have just created.
+//!
+//! Additionally, if you want to create your own protocol, there are some additional concepts:
+//!
+//! - The `StreamMuxer` trait is implemented on structs that hold a connection to a remote and can
+//!   subdivide this connection into multiple substreams. See the `muxing` module.
```

Comment:
I'd mention where `StreamMuxer` fits in among the three objects mentioned above, something like "Muxers are used with handlers, which are used with `Swarm`" (but put better)?